README.md
# Simple Fontellico Progress Dialog ⌛

This package helps to render a progress dialog window with different types of loading indicator.


### New Features 💥
* Internal improvements.
* Optional parameters on constructor.
* Fixing bugs
* Updating example

## Instalation
Include `simple_fontellico_progress_dialog` in your `pubspec.yaml` file:

```yaml
dependencies:
  flutter:
    sdk: flutter
  simple_fontellico_progress_dialog: version
```
### Demo Vertical Mode
<img src="https://raw.githubusercontent.com/ajomuch92/simple-fontelico-progress-dialog-flutter/master/assets/demo.gif" width="200" height="429"/>

### Demo Horizontal Mode
<img src="https://raw.githubusercontent.com/ajomuch92/simple-fontelico-progress-dialog-flutter/master/assets/demo-horizontal.gif" width="200" height="429"/>

## Usage

To use this package, just import it into your file, create a new instance, pass the context as parameter(required).

```dart
import 'package:simple_fontellico_progress_dialog/simple_fontico_loading.dart';

...

SimpleFontelicoProgressDialog _dialog = SimpleFontelicoProgressDialog(context: context);
...

void showDialog() async{
  _dialog.show(message: 'Loading...', type: type);
  await Future.delayed(Duration(seconds: 1));
  _dialog.hide();
}

...

```

### Constructor
|  Name | Description   | Required   | Default   |
| ------------ | ------------ | ------------ | ------------ |
| context  | Buildcontext to render the dialog | True   |   |
| barrierDimisable  | Boolean value to indicate barrierDimisable on dialog |  True  |   |

To show a dialog, use the show method, to hide it, use the hide method. Simple like that.
### Properties used on show method

|  Name | Description   | Required   | Default   |
| ------------ | ------------ | ------------ | ------------ |
| message  | String to set a message into the dialog | True   |   |
| type  | Simple dialog type (normal, threeline, multiline, refresh, hurricane, phoenix, iphone) |  False  |  normal |
| width  | Double value to set the dialog width | False   | 100 |
| height  | Double value to set the dialog height | False   | 120  |
| elevation  | Double value to set the dialog elevation| False   |  5.0 |
| radius  | Double value to set the dialog border radius | False   |  5.0 |
| backgroundColor  | Color value to set the dialog background color | False   | Colors.white  |
| indicatorColor  | Color value to set the indicator color | False   | Colors.blue[600] |
| duration  | Duration value to animate loading indicator | False   | 1000 miliseconds  |
| horizontal  | Boolean value to set if loading has to show on horizontal | False   | false |
| separation  | Double value to set the separation between loading and text | False   | 10.0 |
| textStyle  | Style to customize the text inside dialog | False   | TextStyle(fontSize: 14) |
| textAlign  | Alignment of the message | False   | TextAlign.center |
| hideText  | Boolean value to hide the text widget | False   | false |
| loadingIndicator  | Widget to use when type is set on custom.  | False (Required when type is custom)  |  |

### Properties used on updateMessageText method

|  Name | Description   | Required   | Default   |
| ------------ | ------------ | ------------ | ------------ |
| message  | String to update the message inside an opened dialgo | True   |   |


CHANGELOG.md
<!-- Date format: DD/MM/YYYY -->
## [0.0.1] - First release

## [0.1.0] - Second release

## [0.1.1] - Third release (28/05/2021)

## [0.1.2] - 4th release (04/06/2021)

## [0.1.3] - 4th release (06/06/2021)

## [0.1.4] - 4th release (26/06/2021)

## [0.1.5] - 4th release (26/06/2021)

## [0.2.0] - 5th release (30/06/2021)

## [0.2.1] - 5th release (04/11/2021)

## [0.3.0] - 6th release (24/12/2022)

test/simple_fontellico_progress_dialog_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:simple_fontellico_progress_dialog/simple_fontellico_progress_dialog.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/rotate_icon.dart
import 'package:flutter/material.dart';

/// Class use to render an animated icon with rotation
class RotateIcon extends StatefulWidget {
  final Widget? child;
  final Duration? duration;
  RotateIcon({Key? key, this.child, this.duration}) : super(key: key);

  @override
  _RotateIconState createState() =>
      _RotateIconState(child: child, duration: duration);
}

class _RotateIconState extends State<RotateIcon>
    with SingleTickerProviderStateMixin {
  /// Controller to animate the children
  late AnimationController _animationController;

  /// Widget (icon) to show into the dialog
  final Widget? child;

  /// Duration for the animation
  final Duration? duration;

  _RotateIconState({this.child, this.duration = const Duration(seconds: 1)});

  @override
  void initState() {
    super.initState();
    if (mounted) {
      _animationController = new AnimationController(
        vsync: this,
        duration: duration,
      );
      _animationController.repeat();
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animationController,
      child: child,
      builder: (BuildContext context, Widget? _widget) {
        return new Transform.rotate(
          angle: _animationController.value * 6.3,
          child: _widget,
        );
      },
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_fontellico_progress_dialog/simple_fontico_loading.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Simple Fontellico Progress Dialog Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  SimpleFontelicoProgressDialog? _dialog;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Simple Fontellico Progress Dialog Demo'),
      ),
      body: Center(
        child: GridView.count(
          primary: false,
          padding: const EdgeInsets.all(20),
          crossAxisSpacing: 8,
          mainAxisSpacing: 8,
          crossAxisCount: 2,
          children: <Widget>[
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.normal, 'Normal');
              }, 
              child: const Text('Normal'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.normal, 'Normal Vertical');
              }, 
              child: const Text('Normal Vertical'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.normal, 'Updating');
              }, 
              child: const Text('Updating'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.threelines, 'Three Lines');
              }, 
              child: const Text('Three Lines'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.multilines, 'Multilines');
              }, 
              child: const Text('Multilines'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.refresh, 'Refresh');
              }, 
              child: const Text('Refresh'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.phoenix, 'Phoenix');
              }, 
              child: const Text('Phoenix'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.hurricane, 'Hurricane');
              }, 
              child: const Text('Hurricane'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.iphone, 'Iphone');
              }, 
              child: const Text('Iphone'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.custom, 'Custom');
              }, 
              child: const Text('Custom'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.spinner, 'Spinner');
              }, 
              child: const Text('Spinner'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.cog, 'Cog');
              }, 
              child: const Text('Cog'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.redo, 'Redo');
              }, 
              child: const Text('Redo'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.notch, 'Notch');
              }, 
              child: const Text('Notch'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.bullets, 'Bullets');
              }, 
              child: const Text('Bullets'),
            ),
            ElevatedButton(
              onPressed: (){
                _showDialog(context, SimpleFontelicoProgressDialogType.multiHurricane, 'MultiHurricane');
              }, 
              child: const Text('MultiHurricane'),
            ),
            const SizedBox(height: 10.0,),
          ],
        ),
      ),
       // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  void _showDialog(BuildContext context, SimpleFontelicoProgressDialogType type, String text) async{
    _dialog ??= SimpleFontelicoProgressDialog(context: context);
    if(type == SimpleFontelicoProgressDialogType.custom) {
      _dialog!.show(message: text, type: type, width: 150.0, height: 75.0, loadingIndicator: const Text('C', style: TextStyle(fontSize: 24.0),));
    } else if (text == 'Normal Vertical') {
      _dialog!.show(message: text, type: SimpleFontelicoProgressDialogType.normal, horizontal: false, width: 150.0, height: 75.0, hideText: false, indicatorColor: Colors.red);
    } else if (text == 'Updating') {
      _dialog!.show(message: text, type: SimpleFontelicoProgressDialogType.normal, horizontal: true, width: 150.0, height: 75.0, hideText: false, indicatorColor: Colors.red);
      await Future.delayed(const Duration(seconds: 1));
      _dialog!.updateMessageText('Changing text');
    } else {
      _dialog!.show(message: text, type: type, horizontal: true, width: 150.0, height: 75.0, hideText: false, indicatorColor: Colors.red);
    }
    await Future.delayed(const Duration(seconds: 1));
    _dialog!.hide();
  }
}

lib/fontelico_icons.dart
import 'package:flutter/widgets.dart';

class Fontelico {
  Fontelico._();

  static const _kFontFam = 'Fontelico';
  static const String? _kFontPkg = 'simple_fontellico_progress_dialog';

  /// IconData for bullets icon
  static const IconData bullets =
      IconData(0xe800, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for multiHurricane icon
  static const IconData multiHurricane =
      IconData(0xe983, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for notch icon
  static const IconData notch =
      IconData(0xe801, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for cog icon
  static const IconData cog =
      IconData(0xe802, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for redo icon
  static const IconData redo =
      IconData(0xe803, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for progress_0 icon
  static const IconData progress_0 =
      IconData(0xe804, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for progress_1 icon
  static const IconData progress_1 =
      IconData(0xe805, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for progress_2 icon
  static const IconData progress_2 =
      IconData(0xe806, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for progress_3 icon
  static const IconData progress_3 =
      IconData(0xe807, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for threelines icon
  static const IconData threelines =
      IconData(0xe830, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for multilines icon
  static const IconData multilines =
      IconData(0xe831, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for refresh icon
  static const IconData refresh =
      IconData(0xe832, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for hurricane icon
  static const IconData hurricane =
      IconData(0xe834, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for iphone icon
  static const IconData iphone =
      IconData(0xe838, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for phoenix icon
  static const IconData phoenix =
      IconData(0xe839, fontFamily: _kFontFam, fontPackage: _kFontPkg);

  /// IconData for spinner icon
  static const IconData spinner =
      IconData(0xf110, fontFamily: _kFontFam, fontPackage: _kFontPkg);
}


lib/simple_fontico_loading.dart
library simple_fontellico_progress_dialog;

import 'package:flutter/material.dart';
import 'package:simple_fontellico_progress_dialog/fontelico_icons.dart';
import 'package:simple_fontellico_progress_dialog/rotate_icon.dart';

enum SimpleFontelicoProgressDialogType {
  normal,
  threelines,
  multilines,
  refresh,
  hurricane,
  iphone,
  phoenix,
  notch,
  spinner,
  cog,
  redo,
  bullets,
  multiHurricane,
  custom
}

class SimpleFontelicoProgressDialog {
  /// Value to indicate if dialog is open
  bool _isOpen = false;

  /// String value to set the main message inside the alert dialog
  String message = '';

  /// State Setter used to update the message inside the alert dialog
  late StateSetter setState;

  /// Context to render the dialog
  final BuildContext context;

  /// Bool value to indicate the barrierDismisable of the dialog
  final bool barrierDimisable;

  /// Duration for animation
  final Duration? duration;

  /// Widget indicator when custom is selected
  Widget? _customLoadingIndicator;

  /// Color value to set the indicator color
  Color? _indicatorColor;

  SimpleFontelicoProgressDialog(
      {required this.context,
      this.barrierDimisable = false,
      this.duration = const Duration(milliseconds: 1000)});

  /// Method to render the widget into the dialog
  Widget _getLoadingIndicator(SimpleFontelicoProgressDialogType type) {
    Key key = UniqueKey();
    switch (type) {
      case SimpleFontelicoProgressDialogType.threelines:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.threelines,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.multilines:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.multilines,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.refresh:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.refresh,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.hurricane:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.hurricane,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.iphone:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.iphone,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.phoenix:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.phoenix,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.notch:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.notch,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.spinner:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.spinner,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.cog:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.cog,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.redo:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.redo,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.bullets:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.bullets,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.multiHurricane:
        return RotateIcon(
          key: key,
          duration: this.duration,
          child: Icon(
            Fontelico.multiHurricane,
            color: _indicatorColor,
            size: 40.0,
          ),
        );
      case SimpleFontelicoProgressDialogType.custom:
        return _customLoadingIndicator!;
      case SimpleFontelicoProgressDialogType.normal:
      default:
        return CircularProgressIndicator(
            valueColor: AlwaysStoppedAnimation<Color>(_indicatorColor!));
    }
  }

  /// message: String to indicate a message into the dialog. Required
  /// type: Simple dialog type (normal, threeline, multiline, refresh, hurricane, phoenix, iphone)
  /// height: Double value to set the dialog height
  /// width: Double value to set the dialog width
  /// radius: Double value to set the dialog border radius
  /// elevation: Double value to set the dialog elevation
  /// backgroundColor: Color value to set the dialog background color
  /// indicatorColor: Color value to set the indicator color
  /// horizontal: Boolean value to set if loading has to show on horizontal
  /// separation: Double value to set the separation between loading and text
  /// textStyle: Style to customize the text inside dialog
  /// hideText: Boolean value to hide the text widget
  /// loadingIndicator: Widget to use when type is custom
  /// textAlign: Value to align the text
  void show(
      {required String message,
      SimpleFontelicoProgressDialogType type =
          SimpleFontelicoProgressDialogType.normal,
      double height = 100,
      double width = 120,
      double radius = 5.0,
      double elevation = 5.0,
      Color backgroundColor = Colors.white,
      Color? indicatorColor,
      bool horizontal = false,
      double separation = 10.0,
      TextStyle textStyle = const TextStyle(fontSize: 14),
      TextAlign textAlign = TextAlign.center,
      bool hideText = false,
      Widget? loadingIndicator}) {
    _indicatorColor = indicatorColor ?? Colors.blue[600];
    if (type == SimpleFontelicoProgressDialogType.custom) {
      assert(loadingIndicator != null,
          'Loading indicator must not be null when is custom');
      _customLoadingIndicator = loadingIndicator;
    }
    _isOpen = true;
    this.message = message;
    showDialog(
        context: context,
        barrierDismissible: barrierDimisable,
        useSafeArea: true,
        builder: (BuildContext context) {
          return WillPopScope(
            onWillPop: () => Future.value(barrierDimisable),
            child: Dialog(
              backgroundColor: Colors.transparent,
              insetPadding: EdgeInsets.all(0.0),
              child: StatefulBuilder(builder: (context, setState) {
                this.setState = setState;
                return Center(
                  child: Container(
                    height: height,
                    width: width,
                    decoration: BoxDecoration(
                        color: backgroundColor,
                        borderRadius:
                            BorderRadius.all(Radius.circular(radius))),
                    child: !horizontal
                        ? Column(
                            mainAxisAlignment: MainAxisAlignment.center,
                            crossAxisAlignment: CrossAxisAlignment.center,
                            mainAxisSize: MainAxisSize.min,
                            children: _getChildren(type, horizontal, separation,
                                textStyle, textAlign, hideText),
                          )
                        : Row(
                            mainAxisAlignment: MainAxisAlignment.center,
                            crossAxisAlignment: CrossAxisAlignment.center,
                            mainAxisSize: MainAxisSize.min,
                            children: _getChildren(type, horizontal, separation,
                                textStyle, textAlign, hideText),
                          ),
                  ),
                );
              }),
            ),
          );
        });
  }

  /// Method to hide the dialog
  void hide() {
    if (_isOpen) {
      Navigator.of(context).pop();
      _isOpen = false;
    }
  }

  /// Method to update the message text when dialog is open
  void updateMessageText(String message) {
    if (_isOpen) {
      setState(() {
        this.message = message;
      });
    }
  }

  /// Method to get the children inside the dialog
  List<Widget> _getChildren(
      SimpleFontelicoProgressDialogType type,
      bool horizontal,
      double separation,
      TextStyle textStyle,
      TextAlign textAlign,
      bool hideText) {
    if (hideText) {
      return [_getLoadingIndicator(type)];
    }
    return [
      _getLoadingIndicator(type),
      !horizontal
          ? SizedBox(
              height: separation,
            )
          : SizedBox(
              width: separation,
            ),
      Text(
        message,
        style: textStyle,
        textAlign: textAlign,
      )
    ];
  }
}
